cloud-init: add basic support for write_files

mguegan opened the pull request:
This patch will handle the cloud-init `write_files` function.
Currently, only `path`, `permissions`, `owner` and `content` directives are supported.

Tested OK on OpenBSD 6.6/amd64

Commit:
cloud-init: add basic support for write_files

 - path, permissions, owner and content are supported

pyr commented on cloud-init.pl:
```diff
@@ -84,6 +84,25 @@ sub apply_user_data {
+
+  if (defined($data->{write_files})) {
+    foreach my $item (@{ $data->{write_files} }) {
+      open my $fh, ">", $item->{path};
```
It might be worth doing an `mkdir -p` on the path's dirname before attempting to write.
 Something along the lines of:

```perl
use File::Basename;
use File::Path;

[...]

mkpath [dirname($item)], 1, 0755;
```

mguegan replied:
Agreed, will test & update the PR

pyr commented on cloud-init.pl:
```diff
@@ -84,6 +84,25 @@ sub apply_user_data {
+  if (defined($data->{write_files})) {
+    foreach my $item (@{ $data->{write_files} }) {
+      open my $fh, ">", $item->{path};
+      printf $fh "%s", $item->{content};
```
Superfluous `printf`

```suggestion
      print $fh, $item->{content};
```

mguegan replied:
huh, this one will print `$fh` and `$item->{content}` ?

mguegan replied:
Ok I get it : remove the `,`

pyr replied:
Sorry, I typoed

pyr commented on cloud-init.pl:
```diff
@@ -84,6 +84,25 @@ sub apply_user_data {
+      open my $fh, ">", $item->{path};
+      printf $fh "%s", $item->{content};
+      if (defined($item->{permissions})) {
+        my $perms = oct($item->{permissions});
```
To handle both dirs and files, a future improvement could be to check for a `umask` key here and apply that before the `mkpath` and `open` calls, I wouldn't make that a prerequisite for this.

mguegan commented:
I did a mistake by adding a `system()` at the end of the function, I'll correct the whole PR after additional tests.

Commit:
cloud-init: remove system call from write_files and change execution order

`write_files` should not execute file after writing : this should be handled
by `run_cmd` for example. Changing the execution order of `run_cmd` and
`write_files` permits the above workflow.

This commit also add some small corrections :
 - mkpath() is executed before file creation
 - use print instead of printf

pyr reviewed (approved):
Thanks a lot @mguegan!

mguegan commented:
@pyr you've approved the PR, but did you forget to merge it ?

pyr commented:
@mguegan I did, it's all good now
